### ISSUE TITLE
Encoding large images

### ISSUE THREAD

**Issue by KjellRS** [Hhhhhhao/continuous_tokenizer#1]
Looks like impressive performance on Imagenet, but one of the huge advantages of most VAEs is that they can do tiled encoding of images of arbitrary size (modulo the downscaling factor). As I understand it the 1D latents in SoftVQ-VAE don't correspond to a spatial location, this means that for something like FFHQ (1024x1024) you get either huge tiles of 32/64 tokens with obvious seams or very few tokens. And using it to encode other aspect ratios in datasets like LAION would need a lot of work in pre/postprocessing.

Would it be feasible to train the network - assuming 64 tokens so we can put them on an 8x8 grid - to have spatial meaning after all or would that ruin the performance? This is what I'm thinking:

1. Take a 512x512 image crop.
2. Encode each 256x256 quadrant so you end up with 2x2x64 tokens
3. Rearrange the tokens as (2*8)x(2*8) = 16x16 tokens
4. Pick random x and y offsets from 0-7, let's say (3,4)
5. Take an 8x8 crop of tokens with (3,4) as the top left corner and decode
6. Compare to the corresponding 256x256 image crop using the existing losses

This should guide the network to put the relevant information for each 256/8 = f/32 patch in the one latent corresponding to that grid square. Then you can use it on arbitary sizes, 256x256 -> 8x8. 512x512 -> 16x16. 1024x1024 -> 32x32. 1280x768 -> 40x24. You can probably even interpolate/geomean overlapping tokens instead of tiling for even smoother seams. What do you think? I don't have the horsepower to test but I have the feeling that you do.

**Comment by Hhhhhhao** [comment]
Thanks for your interests.
An easier way to incorporate arbitrary resolution might be just training with mixed resolution images with 64/32 tokens. 
We showed in our paper it works separately for 512 and 256 images, so it would be viable for mixed resolution too. 
We also released the 512 models in case you want to have a try.

**Comment by KjellRS** [comment]
So I think I found the way for HQ images. Here's an image (pretty much worst case) from VITON-HD:
![00016_00](https://github.com/user-attachments/assets/561e75c8-a014-44c5-a7f3-fc8074064004)
There's no way I'm training on 768x1024 images directly and besides 64x32 is probably too small a latent anyway.

Tiling with the pretrained "SoftVQVAE/softvq-b-64" checkpoint performs quite poorly, not that surprising:
![00016_00 rec](https://github.com/user-attachments/assets/6d03add7-8092-4607-8a13-114791b777ad)

Finetuning helps quite a bit but you still get bad corners, bad gradients and broken lines:
![exp102_softvq-b-64_00016_00 rec](https://github.com/user-attachments/assets/b2479bcf-32ae-490f-b65b-c3e4df10f6ea)

However, I trained the network to also be able to decode half-overlapping patches and this fixes almost everything:
![exp102_softvq-b-64_00016_00 overlap 0045000](https://github.com/user-attachments/assets/35aaec9a-7478-49a1-81b1-d2c490d6e6f0)

Here's an unfolded view of the same output with all the decoded patches:
![exp102_softvq-b-64_00016_00 unfolded 0045000](https://github.com/user-attachments/assets/747d0cec-ced2-42d2-98cd-6d6a8dd5b92a)

These are the changes to the tokenizer itself:
```
    def forward(self, input, x2 = False):
        if x2:
            return self.forward_x2(input)
        else:
            quant, codebook_loss, info = self.encode(input)
            recons_imgs = self.decode(quant)
            return input, recons_imgs, codebook_loss, info
```
```
    def get_offsets(self, offsets, l_h, l_w):
        offsets = []
        top_h = torch.randint(high=2, size=()).item() * (l_h // 2)
        top_w = torch.randint(high=2, size=()).item() * (l_w // 2)
        for offset_h in [top_h,top_h + (l_h // 2)]:
            for offset_w in [top_w,top_w + (l_w // 2)]:
                offsets.append((offset_h, offset_w))        
```
```    
    def forward_x2(self, x_2x2):
        # L = 64 as 8x8
        l_h = 8
        l_w = 8
        offsets = self.get_offsets(offsets, l_h, l_w)
        b, c, i_h, i_w = x_2x2.shape
        # 2x2 tiles
        t_h = 2
        t_w = 2
        # Each tile/patch is 256x256
        p_h = i_h // t_h
        p_w = i_w // t_w
        # Downscale factors
        d_h = p_h // l_h
        d_w = p_h // l_w
        # Split and encode patches separately
        patches = rearrange(x_2x2, 'b c (t_h p_h) (t_w p_w) -> (b t_h t_w) c p_h p_w', t_h=t_h, t_w=t_w)
        z, diff, info = self.encode(patches)
        z = rearrange(z, 'b (l_h l_w) d -> b l_h l_w d', l_h=l_h, l_w=l_w)
        z = rearrange(z, '(b t_h t_w) l_h l_w d -> b (t_h l_h) (t_w l_w) d', t_h=t_h, t_w=t_w)        
        # Get patches of super-image and corresponding patches of super-z
        imgs_out = []
        zs_out = []
        for idx in range(b):
            for offset_h, offset_w in offsets:
                x = x_2x2[idx,:,d_h*offset_h:d_h*(offset_h+l_h), d_w*offset_w:d_w*(offset_w+l_w)]                
                imgs_out.append(x)
                z_out = z[idx,offset_h:offset_h+l_h,offset_w:offset_w+l_w,:]     
                zs_out.append(z_out)
        imgs_out = torch.stack(imgs_out)
        zs_out = torch.stack(zs_out)
        zs_out = rearrange(zs_out, 'b l_h l_w d -> b (l_h l_w) d')
        # Quantize and decode as normal patches
        rec_out = self.decode(zs_out)
        return imgs_out, rec_out, diff, info
```
Training with one 512x512 input is equal to training with four 256x256 inputs so I reduce loader batch size to maintain the global batch size/memory consumption etc.

Some utility functions:
```
def encode(super_img, t_h, t_w, vq_model, l_h, l_w):
    patch_img = rearrange(super_img, 'b c (t_h p_h) (t_w p_w) -> (b t_h t_w) c p_h p_w', t_h=t_h, t_w=t_w)
    patch_z, _, _ = vq_model.encode(patch_img)
    patch_z = rearrange(patch_z, 'b (l_h l_w) d -> b l_h l_w d', l_h=l_h, l_w=l_w)
    super_z = rearrange(patch_z, '(b t_h t_w) l_h l_w d -> b (t_h l_h) (t_w l_w) d', t_h=t_h, t_w=t_w)
    return super_z
```

```
def decode(super_z, t_h, t_w, vq_model, p_h, p_w, l_h, l_w, overlap = True):
    offsets = []
    for offset_h in range(0, t_h*l_h - l_h // 2, l_h // 2):
        for offset_w in range(0, t_w*l_w - l_w // 2, l_w // 2):
            top_edge = offset_h == 0
            bottom_edge = offset_h == (t_h-1)*l_h
            left_edge = offset_w == 0
            right_edge = offset_w == (t_w-1)*l_w
            offsets.append((offset_h, offset_w, top_edge, bottom_edge, left_edge, right_edge))

    patch_z = []
    for offset_h, offset_w, _, _, _, _ in offsets:
        tile_z = super_z[:, offset_h:offset_h+l_h, offset_w:offset_w+l_w]
        patch_z.append(tile_z)

    patch_z = torch.cat(patch_z)
    patch_z = rearrange(patch_z, 'b l_h l_w d -> b (l_h l_w) d')
    with torch.no_grad():
        patch_img = vq_model.decode(patch_z)
    if not overlap:
        img_unfolded = rearrange(patch_img, '(b t_h t_w) c p_h p_w -> b c (t_h p_h) (t_w p_w)', t_h=2*t_h-1, t_w=2*t_w-1)
        return img_unfolded        
    img_overlap = torch.zeros(3, t_h*p_h, t_w*p_w)
    for patch_tile, (offset_h, offset_w, top_edge, bottom_edge, left_edge, right_edge) in zip(patch_img, offsets):
        img_offset_h = offset_h*(p_h // l_h)
        img_offset_w = offset_w*(p_w // l_w)
        if top_edge:
            patch_begin_h = 0
        else:
            patch_begin_h = p_h // 4
        if bottom_edge:
            patch_end_h = p_h
        else:
            patch_end_h = p_h * 3 // 4
        if left_edge:
            patch_begin_w = 0
        else:
            patch_begin_w = p_w // 4
        if right_edge:
            patch_end_w = p_w
        else:
            patch_end_w = p_w * 3 // 4
        patch_cutout = patch_tile[:, patch_begin_h:patch_end_h, patch_begin_w:patch_end_w]
        img_overlap[:, img_offset_h+patch_begin_h:img_offset_h+patch_end_h, img_offset_w+patch_begin_w:img_offset_w+patch_end_w] = patch_cutout
    return img_overlap
```

And then finally the calling code:
```
        t_w = 3
        t_h = 4
        p_h = 256
        p_w = 256
        l_h = 8
        l_w = 8        
        img = transform(img).unsqueeze(0).to(device)
        super_z = encode(img, t_h, t_w, vq_model, l_h, l_w)
        img_overlap = decode(super_z, t_h, t_w, vq_model, p_h, p_w, l_h, l_w)        
        img_unfolded = decode(super_z, t_h, t_w, vq_model, p_h, p_w, l_h, l_w, overlap = False)
```

Wow, that got to be a long comment... I should probably make this a PR or show it in a fork or something.

**Comment by Hhhhhhao** [comment]
Thanks for your exploration.
The results looks very interesting. We can open a PR for this, adding condition determining the input size. And if it is larger than the training size, we can use the above method to do reconstruction/generation.

**Comment by KjellRS** [comment]
As long as the network is trained/finetuned with double size input images, yes. If it's not it will look like this:
![00016_00 unfolded](https://github.com/user-attachments/assets/bd5dab69-a2c7-4109-b811-2685777d64fd)
Fortunately it very quickly adapted from your pretrained checkpoint so it is not necessary to start from scratch, I don't know if that affects the final quality though. I did start a full training and it seemed to start out okay but I abandoned it because it'd take so long to complete for probably no gain.

**Comment by Hhhhhhao** [comment]
Thanks for the exploration. If you can upload the weights on huggingface of your finetuned model we can make a notebook demo for larger resolution together. I can share my huggingface uploading code if you need. Let me know.

**Comment by KjellRS** [comment]
I added the checkpoint here:
https://huggingface.co/kjellrs/SoftVQVAE/tree/main

It's a softvq-b-64 checkpoint, you can load it on top of your model like this:
```
    vq_model = SoftVQModel.from_pretrained('SoftVQVAE/softvq-b-64')
    ckpt_path = f'your/path/here/078000.pt'
    ckpt = torch.load(ckpt_path, map_location="cpu")
    use_ema = False
    if use_ema:
        state_dict = ckpt['ema']
    else:
        state_dict = ckpt['model']
    vq_model.load_state_dict(state_dict, strict = False)
```
Note that this model has been finetuned on a set of human images, it may have lost some of the ability to reproduce other imagenet categories.
